### PR TITLE
[NativeAOT/ARM] Bail out on IMAGE_REL_BASED_THUMB_BRANCH24 with >24-bit addends

### LINKED ISSUE #97750: [NativeAOT/ARM] Assert failure in PutThumb2BlRel24
Crashes when compiling System.Net.Quic.Functional.Tests library tests:

```
[/home/navara/runtime/src/libraries/System.Net.Quic/tests/FunctionalTests/System.Net.Quic.Functional.Tests.csproj::TargetFramework=net9.0-linux]
  Process terminated. Assertion failed.
     at ILCompiler.DependencyAnalysis.Relocation.PutThumb2BlRel24(UInt16* p, Int32 imm24) in /home/navara/runtime/src/coreclr/tools/Common/Compiler/DependencyAnalysis/Relocation.cs:line 177
     at ILCompiler.DependencyAnalysis.Relocation.WriteValue(RelocType relocType, Void* location, Int64 value) in /home/navara/runtime/src/coreclr/tools/Common/Compiler/DependencyAnalysis/Relocation.cs:line 499
     at ILCompiler.ObjectWriter.ObjectWriter.EmitOrResolveRelocation(Int32 sectionIndex, Int64 offset, Span`1 data, RelocType relocType, String symbolName, Int64 addend) in /home/navara/runtime/src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/ObjectWriter/ObjectWriter.cs:line 185
     at ILCompiler.ObjectWriter.ObjectWriter.EmitObject(String objectFilePath, IReadOnlyCollection`1 nodes, IObjectDumper dumper, Logger logger) in /home/navara/runtime/src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/ObjectWriter/ObjectWriter.cs:line 443
     at ILCompiler.ObjectWriter.ObjectWriter.EmitObject(String objectFilePath, IReadOnlyCollection`1 nodes, NodeFactory factory, ObjectWritingOptions options, IObjectDumper dumper, Logger logger) in /home/navara/runtime/src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/ObjectWriter/ObjectWriter.cs:line 533
     at ILCompiler.RyuJitCompilation.CompileInternal(String outputFile, ObjectDumper dumper) in /home/navara/runtime/src/coreclr/tools/aot/ILCompiler.RyuJit/Compiler/RyuJitCompilation.cs:line 111
     at ILCompiler.Compilation.ILCompiler.ICompilation.Compile(String outputFile, ObjectDumper dumper) in /home/navara/runtime/src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/Compilation.cs:line 528
     at ILCompiler.Program.Run() in /home/navara/runtime/src/coreclr/tools/aot/ILCompiler/Program.cs:line 585
     at ILCompiler.ILCompilerRootCommand.<>c__DisplayClass236_0.<.ctor>b__0(ParseResult result) in /home/navara/runtime/src/coreclr/tools/aot/ILCompiler/ILCompilerRootCommand.cs:line 292
     at System.CommandLine.Invocation.InvocationPipeline.Invoke(ParseResult parseResult)
     at System.CommandLine.ParseResult.Invoke()
     at ILCompiler.Program.Main(String[] args) in /home/navara/runtime/src/coreclr/tools/aot/ILCompiler/Program.cs:line 753
  Aborted
/home/navara/runtime/artifacts/bin/coreclr/linux.arm.Debug/build/Microsoft.NETCore.Native.targets(310,5): error MSB3073: The command ""/home/navara/runtime/artifacts/bin/coreclr/linux.arm.Debug/x64/ilc/ilc" @"/home/navara/runtime/artifacts/obj/System.Net.Http.Functional.Tests/Debug/net9.0-linux/native/System.Net.Http.Functional.Tests.ilc.rsp"" exited with code 134. 
```

### PR DESCRIPTION
**Pull request by filipnavara**
... and let the linker generate thunks.

Fixes #97750

### CODE REVIEW

File: src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/ObjectWriter/ObjectWriter.cs [hunk 1]
```diff
@@ -169,20 +169,29 @@ private protected static ObjectNodeSection GetSharedSection(ObjectNodeSection se
...
-                    addend += Relocation.ReadValue(relocType, (void*)pData);
-                    addend |= definedSymbol.Value & maskThumbBitIn;
-                    addend -= offset;
-                    Relocation.WriteValue(relocType, (void*)pData, addend);
+                    adjustedAddend += definedSymbol.Value & ~maskThumbBitOut;
+                    adjustedAddend += Relocation.ReadValue(relocType, (void*)pData);
+                    adjustedAddend |= definedSymbol.Value & maskThumbBitIn;
+                    adjustedAddend -= offset;
+
+                    if (relocType is IMAGE_REL_BASED_THUMB_BRANCH24 && (uint)adjustedAddend >= 0xFFFFFF)
```
Comment (filipnavara) [inline src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/ObjectWriter/ObjectWriter.cs:187]: On second thought... I am not sure this is the correct check for the signed values.

Comment (filipnavara) [reply src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/ObjectWriter/ObjectWriter.cs:187]: Perhaps I can make the `Relocation.FitsInThumb2BlRel24` public and call it here.

**Comment by jkotas** [comment]
> let the linker generate thunks.

Is the linker guaranteed to find space to generate the thunks? How reliable is this fix?

**Comment by filipnavara** [comment]
> Is the linker guaranteed to find space to generate the thunks? How reliable is this fix?

It is not guaranteed to find a space but it should be a hard fail if it does not. In practice I expect this should effectively double the size of code before the error is hit again (to `1 << 25` bytes).

We may investigate alternative mitigating strategies but this is the cheap way to increase the size limit.

**Comment by jkotas** [comment]
Are the thunks placed outside the managed code section? If they are part the managed code section, there may be fixes needed in the code manager to ignore them.

**Comment by filipnavara** [comment]
> Are the thunks placed outside the managed code section?

That's a good question. They should not interleave with existing code and I assume they would have to be placed at the end of the same section as the code. That means they will not have reliable unwinding data, and that could be a problem (especially if it picks unwinding code from last method).

I'll put this back to draft for now and verify the assumptions later.

**Comment by filipnavara** [comment]
> Are the thunks placed outside the managed code section?

The linker creates a separate `.text.thunks` section, so it should not interfere. The hijack signal handler can catch the thread inside the thunk but it will bail out since it's not in a managed code at that point.

**Review by jkotas — APPROVED** [review]
Thank you for checking

**Pull request merged**